README.md

```diff
@@ -2,4 +2,61 @@
+
+In addition to operators that have a one-to-one mapping with PromQL constructs, the Volcano model also describes so-called Exchange operators which can be used for flow control and optimizations, such as concurrency or batched selects. An example of an _Exchange_ operator is described in the [Intra-operator parallelism](#intra-operator-parallelism) section.
+
+## Inter-operator parallelism
+
+Since operators are independent and rely on a common interface for pulling data, they can be run in parallel to each other. As soon as one operator has processed data from an evaluation step, it can pass the result onward so that its upstream can immediatelly start working on it.
+
+<p align="center">
+  <img src="./assets/promql-pipeline.png" />
+</p>
+
+
+## Intra-operator parallelism
+
+Parallelism can also be added within individual operators using a parallel coalesce exchange operator. Such exchange operators are indistinguishable from regular operators to their upstreams since they respect the same `Next()` interface.
+
+<p align="center">
+  <img src="./assets/parallel-coalesce.png" />
+</p>
+
+
+## Memory management 
+
+### Step vector allocations
+
+One challenge with the streamed execution model is knowing how much memory to allocate in each operator for each step. 
+
+To work around this issue, operators expose a `Series()` method which returns the labels for all time series that they will ever produce. Operators at the very bottom of the tree, like vector and matrix selectors, have this information since they are responsible for loading data from storage. Other operators can then call `Series()` on the downstream operator and pre-compute all possible outputs.
+
+Even thought this might look like an expensive operation, its cost is identical to just one evaluation step. Knowing sizes of input and output vectors also allows us to:   
+* Allocate memory very precisely by properly sizing vector pools (see section bellow),
+* Use arrays instead of maps for indexing data and performing lookups, leading to even further memory usage and faster execution time, and
+* Use tight loops in operators by eliminating a large amount of conditional statements.
+
+### Vector pools
+
+Since time series are decoded one step at a time, vectors between execution execution steps can be recycled manually instead of relying on the garbage collector. Each operator has its own pool that it uses to allocate new step vectors and send results to its upstream. Whenever the upstream operator is finished with processing a step vector, it will return that vector to the pool of its downstream so that it can be reused again for subsequent steps.
+
+## Concurrency control
```

Comment:
What about memory control?

Reply:
We don't have anything in place yet. Should we look at something like max samples as a proxy, or try to estimate memory more precisely?

Reply:
As an end-user, "max samples" forces me to guess and experiment to find an acceptable limit that works with the resources I actually have. Personally, I tend to _under_ estimate heavily here, because I don't know precisely how the system will behave - better to have waste than to run out of resources, in my mind.

But that's not really optimal. As an end-user, what I _really_ want to do is say something like "use up to N Mb of memory" or "use no more than 75% of the available system memory." That allows me to maximize my resource usage with confidence (and cuts out the need to guess-and-check over and over again).

If possible, estimating the amount of memory used and expressing that as a limit would be preferable from my perspective.

Reply:
I agree, proxies like max samples and max series can be tricky to configure. Another argument for using memory limits directly is the fact that memory usage depends in part on the implementation itself. Users would somehow have to know that the engine got better or worse at managing memory and adjust max samples accordingly.